tools/cldr-code/src/main/java/org/unicode/cldr/util/CldrUtility.java

```diff
@@ -1573,6 +1574,37 @@ public static String getCldrBaseDirHash() {
      * @return the hash, like "9786e05e95a2e4f02687fa3b84126782f9f698a3"
      */
     public final static String getGitHashForDir(String dir) {
+        // First, try just reading .git/HEAD
+        final File parent = new File(dir);
+        final File gitDir = new File(dir, ".git");
+        final File headfile = new File(gitDir, "HEAD");
+        if (headfile.canRead()) {
+            // Try this first, fallback to git commands
+            try {
+                String s = Files.readString(headfile.toPath());
+                if (s != null && !s.isBlank()) {
+                    s = s.trim();
+                    if(s.startsWith("ref: ")) {
+                        s = s.substring(5); // refs/heads/main
+                        final Path refPath = gitDir.toPath().resolve(s);
+                        if (refPath.startsWith(gitDir.toPath())) {
+                            s = Files.readString(refPath);
+                            if (s != null && !s.isBlank()) {
+                                return s.trim();
+                            }
+                        } else { // ignore something like refs: ../../../yourfiles
+                            System.err.println("Ignoring strange git refPath " + refPath);
+                        }
+                    } // else, maybe detached head
+                    return s.trim();
+                }
+            } catch (IOException e) {
+                System.err.println(e + ": readString failed for " + headfile);
+                e.printStackTrace();
+            }
+
+        }
+
```

Comment:
this method is too long, around 70 lines; most of what's above here should be in one subroutine, and most of what's below here should be in another subroutine

Reply:
rewrote as separate subroutines